### ISSUE TITLE
[gomod] Propagate the effect of the 'goproxy_url' config option to SBOM

### ISSUE THREAD

**Issue by eskultety** [containerbuildsystem/cachi2#577]
If consumers use a private goproxy, then without formatting it to the PURL the generic package locator is not accurate to describe the package location.
The [PURL spec ](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst) sadly doesn't mention anything that would be even remotely related to what we need here with `repository_url/download_url` being the closest having a different meaning.
We'll probably have to come up with a custom qualifier, i.e. `proxy` to denote that a proxy was used to locate the package given its identifier.

**Comment by sswguo** [comment]
@eskultety hi, is it possible that consumer config multiple proxy urls as follows: ? 
`GOPROXY=https://private-proxy.com,https://proxy.golang.org,direct`

If so, do we need to identify the accurate one per go module ?  I tried the command and seems there is no record for that info:
```
GOPROXY=https://proxy.golang.org,direct go mod download -json github.com/gin-gonic/gin@latest        
{
	"Path": "github.com/gin-gonic/gin",
	"Version": "v1.11.0",
	"Query": "latest",
	"Info": "~/go/pkg/mod/cache/download/github.com/gin-gonic/gin/@v/v1.11.0.info",
	"GoMod": "~/go/pkg/mod/cache/download/github.com/gin-gonic/gin/@v/v1.11.0.mod",
	"Zip": "~/go/pkg/mod/cache/download/github.com/gin-gonic/gin/@v/v1.11.0.zip",
	"Dir": "~/go/pkg/mod/github.com/gin-gonic/gin@v1.11.0",
	"Sum": "h1:OW/6PLjyusp2PPXtyxKHU0RbX6I/l28FTdDlae5ueWk=",
	"GoModSum": "h1:+iq/FyxlGzII0KHiBGjuNn4UNENUlKbGlNmc+W50Dls="
```

But if it happens without any of the proxy, there will be an additional field `Origin` in place:
```
GOPROXY=direct go mod download -json github.com/gin-gonic/gin@latest                                                                     🐍 hermeto 13:51:24
{
	"Path": "github.com/gin-gonic/gin",
	"Version": "v1.11.0",
	"Query": "latest",
	"Info": "~/go/pkg/mod/cache/download/github.com/gin-gonic/gin/@v/v1.11.0.info",
	"GoMod": "~/go/pkg/mod/cache/download/github.com/gin-gonic/gin/@v/v1.11.0.mod",
	"Zip": "~/go/pkg/mod/cache/download/github.com/gin-gonic/gin/@v/v1.11.0.zip",
	"Dir": "~/go/pkg/mod/github.com/gin-gonic/gin@v1.11.0",
	"Sum": "h1:OW/6PLjyusp2PPXtyxKHU0RbX6I/l28FTdDlae5ueWk=",
	"GoModSum": "h1:+iq/FyxlGzII0KHiBGjuNn4UNENUlKbGlNmc+W50Dls=",
	"Origin": {
		"VCS": "git",
		"URL": "https://github.com/gin-gonic/gin",
		"Hash": "6ad6205e9c94a4b8a320219e28c37c29d22a7a2c",
		"TagSum": "t1:yK0MyvqFzQnCd/LSHSL150cX+UpEII14IaeQYlJIJJI=",
		"Ref": "refs/tags/v1.11.0"
	}
}
```

**Comment by eskultety** [comment]
@sswguo thank you for diving into this deeper, it might have just turned this into something much more "interesting".

> [@eskultety](https://github.com/eskultety) hi, is it possible that consumer config multiple proxy urls as follows: ? `GOPROXY=https://private-proxy.com,https://proxy.golang.org,direct`
> 

Yes, it is. In fact, the default is `GOPROXY=https://proxy.golang.org,direct` based on https://go.dev/ref/mod#environment-variables

> If so, do we need to identify the accurate one per go module ? I tried the command and seems there is no record for that info:

Correlating your hands-on experience with the docs https://go.dev/ref/mod#environment-variables if we just pass through the variable there's no way of knowing which one was used, it's all Go internals which is completely opaque to us.
What we'd probably have to do in that case is parse the proxies and apply their effect individually, yeah, "ain't nobody got time for that" but more importantly there's just no benefit.

However, what your finding now means is that in context of https://github.com/hermetoproject/hermeto/pull/1170, more specifically this bit:

> SBOM must be marked to indicate whether a proxy was used or not:

We should pretty much always record it that way for Go packages due to the default `GOPROXY` value.
CC @a-ovchinnikov ^^

> But if it happens without any of the proxy, there will be an additional field `Origin` in place:
> 
> ```
> GOPROXY=direct go mod download -json github.com/gin-gonic/gin@latest                                                                     🐍 hermeto 13:51:24
> {
> 	"Path": "github.com/gin-gonic/gin",
> 	"Version": "v1.11.0",
> 	"Query": "latest",
> 	"Info": "~/go/pkg/mod/cache/download/github.com/gin-gonic/gin/@v/v1.11.0.info",
> 	"GoMod": "~/go/pkg/mod/cache/download/github.com/gin-gonic/gin/@v/v1.11.0.mod",
> 	"Zip": "~/go/pkg/mod/cache/download/github.com/gin-gonic/gin/@v/v1.11.0.zip",
> 	"Dir": "~/go/pkg/mod/github.com/gin-gonic/gin@v1.11.0",
> 	"Sum": "h1:OW/6PLjyusp2PPXtyxKHU0RbX6I/l28FTdDlae5ueWk=",
> 	"GoModSum": "h1:+iq/FyxlGzII0KHiBGjuNn4UNENUlKbGlNmc+W50Dls=",
> 	"Origin": {
> 		"VCS": "git",
> 		"URL": "https://github.com/gin-gonic/gin",
> 		"Hash": "6ad6205e9c94a4b8a320219e28c37c29d22a7a2c",
> 		"TagSum": "t1:yK0MyvqFzQnCd/LSHSL150cX+UpEII14IaeQYlJIJJI=",
> 		"Ref": "refs/tags/v1.11.0"
> 	}
> }
> ```

Yes, again, based on the docs that should always be present IIUC because `direct` translates to "fetch from VCS". Now, again in context of https://github.com/hermetoproject/hermeto/pull/1170  `Origin` would be our only indicator that NO proxy was used in fetching a particular module and hence allowing us to selectively record a `proxy` list for modules that were actually fetched from a proxy and omitting it for those that weren't.

**Comment by eskultety** [comment]
Based on the broader context and a clear connection to some other work in progress I'm inclined to remove the `my first issue` label from this issue.

**Comment by sswguo** [comment]
> Yes, again, based on the docs that should always be present IIUC because `direct` translates to "fetch from VCS". Now, again in context of [#1170](https://github.com/hermetoproject/hermeto/pull/1170) `Origin` would be our only indicator that NO proxy was used in fetching a particular module and hence allowing us to selectively record a `proxy` list for modules that were actually fetched from a proxy and omitting it for those that weren't.

So IIUC we should at least add the `Origin(ParsedOrigin)` into `ParsedModule`, and then to see how to add the proxy list later based on further discussion.

**Comment by sswguo** [comment]
@eskultety @a-ovchinnikov  how about adding the proxy into the properties ? for example the generated SBOM for module as follows:

```
{
    'name': 'rsc.io/sampler',
    'properties': [
        {
            'name': 'hermeto:found_by',
            'value': 'hermeto',
        },
        {
            'name': 'hermeto:go:proxy:config',
            'value': 'https://proxy.golang.org,direct',
        },
    ],
    'purl': 'pkg:golang/rsc.io/sampler@v1.3.0?type=module',
    'type': 'library',
    'version': 'v1.3.0',
},
```

**Comment by eskultety** [comment]
> [@eskultety](https://github.com/eskultety) [@a-ovchinnikov](https://github.com/a-ovchinnikov) how about adding the proxy into the properties ? 

@sswguo the proxy will become a property, a snippet from @a-ovchinnikov's proposal (https://raw.githubusercontent.com/a-ovchinnikov/cachi2/2481b4b372e8880eca0c49c81e3dca594cffa65f/docs/design/proxy-support.md):

> The extra field will become a property for CycloneDX SBOMs and an annotation
for SPDX SBOMs.

so I think a variation of 

```
{
            'name': 'hermeto:go:proxy:config',
            'value': 'https://proxy.golang.org,direct',
        },
```

is planned, however, what you're proposing is tied to Go too much because whoever decides to process the SBOM would have to filter based on `go:proxy:config` whereas we'd want something generic and common for all records, hence the proposal just mentions `proxy` as the keyword.

**Comment by sswguo** [comment]
> hence the proposal just mentions `proxy` as the keyword.

Yeah I see, and I tried to make the PR https://github.com/hermetoproject/hermeto/pull/1179, this is a bit more commits than planned ; -) , let's see if it's compliant to @a-ovchinnikov 's proposal.

**Comment by eskultety** [comment]
@a-ovchinnikov a quick checkup question in context of https://github.com/hermetoproject/hermeto/pull/1575 - how does the PR address or affect this issue, will the issue be closed by completing the PR? In which case we should link this issue to that PR.

**Comment by a-ovchinnikov** [comment]
In https://github.com/hermetoproject/hermeto/pull/1575 an external reference is added to SBOM of type `distribution` with a comment `proxy URL` and URL pointing to a proxy. It does not affect PURL.
